internal/storage/postgres.go

```diff
@@ -0,0 +1,310 @@
+package storage
+
+import (
+	"context"
+	"github.com/jackc/pgx/v4"
+	"github.com/pkg/errors"
+	"github.com/zamibaru89/gophermart/internal/config"
+	"log"
+)
+
+type PostgresStorage struct {
+	Connection *pgx.Conn
+}
+
+func NewPostgresStorage(c config.ServerConfig) (Repo, *pgx.Conn, error) {
+	conn, err := pgx.Connect(context.Background(), c.DSN)
+	if err != nil {
+		log.Println(err)
+	}
+
+	query := `CREATE TABLE IF NOT EXISTS  users(
+    id serial PRIMARY KEY,
+    username VARCHAR(50) NOT NULL UNIQUE,
+    password VARCHAR(50)
+);
+CREATE TABLE IF NOT EXISTS orders (
+				  orderID           VARCHAR(64) UNIQUE,
+				  userID           INT NOT NULL,
+				  state 	  TEXT NOT NULL,
+				  accrual	double precision ,
+					withdraw double precision,
+					uploaded_at TIMESTAMP );
+
+CREATE TABLE IF NOT EXISTS balance (
+				userID           INT UNIQUE NOT NULL,
+				balance double precision);
+
+CREATE TABLE IF NOT EXISTS withdrawals  (
+				userID				INT NOT NULL,
+  				orderID				VARCHAR(64) UNIQUE,
+				sum					double precision,
+    			processed_at		TIMESTAMP);
+
+`
+	_, err = conn.Exec(context.Background(), query)
+	if err != nil {
+		log.Println(err)
+		return &PostgresStorage{Connection: conn}, conn, err
+	}
+
+	return &PostgresStorage{Connection: conn}, conn, nil
+}
+
+func (p *PostgresStorage) RegisterUser(credentials Credentials) error {
+
+	query := `
+		INSERT INTO users(
+		username,
+		password
+		)
+		VALUES($1, $2);
+	`
+	_, err := p.Connection.Exec(context.Background(), query, credentials.Username, credentials.Password)
+	if err != nil {
+		log.Println(err)
+		return err
+
+	}
+	return nil
+}
+
+func (p *PostgresStorage) SelectUser(user Credentials) (Credentials, error) {
```

Comment:
мне большей нравится такой написание
```suggestion
func (p *PostgresStorage) GetUserByID(username string) (Credentials, error) { // можно и SelectUserByID
```
тогда сразу понятно по какому принципу мы вытаскиваем пользователя, без изучения имплементации не догадаешься, что тут только username нужен